flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java

```diff
@@ -118,48 +150,290 @@ public static void destroyHDFS() {
+		testHarness2.setup();
+		testHarness2.open();
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness3 = createRescalingTestSink(outDir, 3, 2, 100);
+		testHarness3.setup();
+		testHarness3.open();
+
+		testHarness1.processElement(new StreamRecord<>("test1", 0L));
+		checkFs(outDir, 1, 0, 0, 0);
+
+		testHarness2.processElement(new StreamRecord<>("test2", 0L));
+		checkFs(outDir, 2, 0, 0, 0);
+
+		testHarness3.processElement(new StreamRecord<>("test3", 0L));
+		testHarness3.processElement(new StreamRecord<>("test4", 0L));
+		checkFs(outDir, 4, 0, 0, 0);
+
+		// intentionally we snapshot them in the reverse order so that the states are shuffled
+		OperatorStateHandles mergedSnapshot = AbstractStreamOperatorTestHarness.repackageState(
+			testHarness3.snapshot(0, 0),
+			testHarness1.snapshot(0, 0),
+			testHarness2.snapshot(0, 0)
+		);
+
+		testHarness1 = createRescalingTestSink(outDir, 2, 0, 100);
+		testHarness1.setup();
+		testHarness1.initializeState(mergedSnapshot);
+		testHarness1.open();
+
+		checkFs(outDir, 1, 0, 3, 3);
+
+		testHarness2 = createRescalingTestSink(outDir, 2, 1, 100);
+		testHarness2.setup();
+		testHarness2.initializeState(mergedSnapshot);
+		testHarness2.open();
+
+		checkFs(outDir, 0, 0, 4, 4);
+	}
+
+	@Test
+	public void testScalingUp() throws Exception {
+		final File outDir = tempFolder.newFolder();
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness1 = createRescalingTestSink(outDir, 2, 0, 100);
+		testHarness1.setup();
+		testHarness1.open();
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness2 = createRescalingTestSink(outDir, 2, 0, 100);
+		testHarness2.setup();
+		testHarness2.open();
+
+		testHarness1.processElement(new StreamRecord<>("test1", 1L));
+		testHarness1.processElement(new StreamRecord<>("test2", 1L));
+
+		checkFs(outDir, 2, 0, 0, 0);
+
+		testHarness2.processElement(new StreamRecord<>("test3", 1L));
+		testHarness2.processElement(new StreamRecord<>("test4", 1L));
+		testHarness2.processElement(new StreamRecord<>("test5", 1L));
+
+		checkFs(outDir, 5, 0, 0, 0);
+
+		// intentionally we snapshot them in the reverse order so that the states are shuffled
+		OperatorStateHandles mergedSnapshot = AbstractStreamOperatorTestHarness.repackageState(
+			testHarness2.snapshot(0, 0),
+			testHarness1.snapshot(0, 0)
+		);
+
+		testHarness1 = createRescalingTestSink(outDir, 3, 0, 100);
+		testHarness1.setup();
+		testHarness1.initializeState(mergedSnapshot);
+		testHarness1.open();
+
+		checkFs(outDir, 2, 0, 3, 3);
+
+		testHarness2 = createRescalingTestSink(outDir, 3, 1, 100);
+		testHarness2.setup();
+		testHarness2.initializeState(mergedSnapshot);
+		testHarness2.open();
+
+		checkFs(outDir, 0, 0, 5, 5);
+
+		OneInputStreamOperatorTestHarness<String, Object> testHarness3 = createRescalingTestSink(outDir, 3, 2, 100);
+		testHarness3.setup();
+		testHarness3.initializeState(mergedSnapshot);
+		testHarness3.open();
+
```

Comment:
let's add a separate check here to make sure that nothing was changed just by the initialization.